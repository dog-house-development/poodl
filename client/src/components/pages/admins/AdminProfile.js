import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import React, { Component } from 'react';
import _ from 'lodash';
import Loading from '../../ui/Loading';
import { DeleteButton } from './../../ui/DeleteButton';
import AdminActions from '../../../actions/adminActions';
import DynamicForm from '../../ui/DynamicForm';
import adminInputs from './adminInputs';
import AuthActions from '../../../actions/authActions';

export class AdminProfile extends Component {
    componentDidMount() {
        this.props.adminActions.get(this.props.match.params.id, _.noop, () => {
            this.props.history.push('/pageNotFound');
        });
    }

    componentDidUpdate(prevProps) {
        if (this.authAdminChanged()) {
            const newAuthAdmin = {
                id: this.props.admin._id,
                firstName: this.props.admin.firstName,
                lastName: this.props.admin.lastName,
                seniorCenterId: this.props.admin.seniorCenterId,
                accessLevel: this.props.admin.accessLevel
            };

            this.props.authActions.setCurrentAdmin(newAuthAdmin);
        }
    }

    authAdminChanged() {
        if (this.props.admin) {
            const { firstName, lastName, _id } = this.props.admin;
            const nameChanged =
                firstName !== this.props.auth.admin.firstName || lastName !== this.props.auth.admin.lastName;
            const idChanged = _id === this.props.auth.admin.id;
            return idChanged && nameChanged;
        }
    }

    getAdminName() {
        return this.props.loading ? (
            <Loading content="" />
        ) : (
            _.get(this.props.admin, 'firstName') + ' ' + _.get(this.props.admin, 'lastName')
        );
    }

    handleDeleteClick = () => {
        this.props.adminActions.delete(this.props.match.params.id, () => this.props.history.push('/admins'));
    };

    editAdmin = (modifiedInputs, onSuccess) => {
        this.props.adminActions.edit(_.get(this.props.admin, '_id'), modifiedInputs, onSuccess);
    };

    getDeleteButton() {
        if (this.props.auth.admin.id !== _.get(this.props.admin, '_id')) {
            return (
                <DeleteButton
                    onConfirm={this.handleDeleteClick}
                    confirmQuestion={`Are you sure you want to delete the admin '${this.getAdminName()}'?`}>
                    Delete Admin
                </DeleteButton>
            );
        }
    }

    render() {
        return (
            <div className="page-container">
                <Link to="/admins" className="button small tertiary">
                    <i className="material-icons">keyboard_backspace</i> Back to all admins
                </Link>
                <div className="page-header">
                    <h1>{this.getAdminName()}</h1>
                    {this.getDeleteButton()}
                </div>
                <DynamicForm
                    inputs={adminInputs}
                    editValues={this.editAdmin}
                    values={this.props.admin}
                    editable={true}
                    loading={this.props.loading}
                    errors={this.props.errors}
                />
            </div>
        );
    }
}

export const mapStateToProps = (state, props) => {
    return {
        admin: state.admins.all[props.match.params.id],
        loading: state.admins.loading,
        errors: state.admins.errors,
        auth: state.auth
    };
};

export const mapDispatchToProps = dispatch => {
    return {
        adminActions: bindActionCreators(AdminActions, dispatch),
        authActions: bindActionCreators(AuthActions, dispatch)
    };
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(AdminProfile);
