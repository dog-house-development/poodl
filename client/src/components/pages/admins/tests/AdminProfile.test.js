import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import _ from 'lodash';
import { Link } from 'react-router-dom';
import { AdminProfile, mapStateToProps, mapDispatchToProps } from '../AdminProfile';

configure({ adapter: new Adapter() });

const match = {
    params: {
        _id: '123' //any id you want to set
    }
};
describe('AdminProfile tests', () => {
    let state, props, wrapper, instance;
    const setInstanceAndWrapper = (_props = {}, _state = {}) => {
        state = _.assign(
            {},
            {
                auth: {
                    admin: {
                        superAdmin: true
                    },
                    errors: {}
                },
                admins: {
                    all: {
                        123: {
                            _id: '123',
                            firstName: 'Moose',
                            lastName: 'Man',
                            email: 'mooseman@nowhere.com',
                            superAdmin: true
                        }
                    },
                    errors: {}
                }
            },
            _state
        );
        props = _.assign({}, { match: { params: { id: 123 } } }, _props);
        wrapper = shallow(
            <AdminProfile
                match={match}
                {..._.assign(
                    {},
                    props,
                    mapStateToProps(state, props),
                    mapDispatchToProps(jasmine.createSpy('dispatch'))
                )}
            />
        );
        instance = wrapper.instance();
    };

    beforeEach(() => {
        setInstanceAndWrapper();
    });

    describe('mapStateToProps', () => {
        it('should map state to props', () => {
            expect(mapStateToProps(state, props)).toEqual({
                admin: {
                    _id: '123',
                    firstName: 'Moose',
                    lastName: 'Man',
                    email: 'mooseman@nowhere.com',
                    superAdmin: true
                },
                currentAdminIsSuper: true,
                errors: {}
            });
        });
    });

    describe('mapDispatchToProps', () => {
        it('should map dispatch to props', () => {
            const dispatch = jest.fn();
            expect(JSON.stringify(mapDispatchToProps(dispatch))).toEqual(JSON.stringify({ adminActions: {} }));
        });
    });

    describe('componentDidMount', () => {
        it('should run without breaking', () => {
            spyOn(instance, 'componentDidMount');
            instance.componentDidMount();
        });
    });

    describe('render', () => {
        it('should render correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });
});
