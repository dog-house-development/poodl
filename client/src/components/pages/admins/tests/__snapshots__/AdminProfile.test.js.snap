// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminProfile tests mapDispatchToProps should map dispatch to props 1`] = `
Object {
  "adminActions": Object {
    "create": [Function],
    "delete": [Function],
    "edit": [Function],
    "filter": [Function],
    "get": [Function],
  },
  "authActions": Object {
    "loginAdmin": [Function],
    "logoutAdmin": [Function],
    "refreshToken": [Function],
    "setCurrentAdmin": [Function],
  },
}
`;

exports[`AdminProfile tests mapStateToProps should map state to props 1`] = `
Object {
  "admin": Object {
    "_id": "123",
    "email": "mooseman@nowhere.com",
    "firstName": "Moose",
    "lastName": "Man",
    "superAdmin": true,
  },
  "auth": Object {
    "admin": Object {
      "superAdmin": true,
    },
    "errors": Object {},
  },
  "errors": Object {},
  "loading": false,
}
`;

exports[`AdminProfile tests render should render correctly 1`] = `
<div
  className="page-container"
>
  <Link
    className="button small tertiary icon"
    replace={false}
    to="/admins"
  >
    <i
      className="material-icons button-icon"
    >
      keyboard_backspace
    </i>
     Back to all admins
  </Link>
  <div
    className="page-header"
  >
    <h1>
      Moose Man
    </h1>
    <div
      className="button-list"
    >
      <OnClickOutside(Dropdown)
        align="right"
        eventTypes={
          Array [
            "mousedown",
            "touchstart",
          ]
        }
        excludeScrollbar={false}
        icon="more_vert"
        kind="tertiary"
        outsideClickIgnoreClass="ignore-react-onclickoutside"
        preventDefault={false}
        stopPropagation={false}
      >
        <ConfirmButton
          className="dropdown-content-row medium"
          message="Are you sure you want to delete the admin 'Moose Man'?"
          onConfirm={[Function]}
          title="Confirm Delete"
        >
          Delete Admin
        </ConfirmButton>
      </OnClickOutside(Dropdown)>
    </div>
  </div>
  <DynamicForm
    editValues={[Function]}
    editable={true}
    errorDescription=""
    errors={Object {}}
    inputs={
      Array [
        Object {
          "id": "personalInfo",
          "inputs": Array [
            Object {
              "id": "name",
              "inputs": Array [
                Object {
                  "id": "firstName",
                  "kind": "field",
                  "label": "First name",
                  "style": Object {
                    "flex": 3,
                  },
                  "type": "text",
                },
                Object {
                  "id": "middleInitial",
                  "kind": "field",
                  "label": "M.I.",
                  "style": Object {
                    "flex": 1,
                  },
                },
                Object {
                  "id": "lastName",
                  "kind": "field",
                  "label": "Last name",
                  "style": Object {
                    "flex": 3,
                  },
                  "type": "text",
                },
                Object {
                  "id": "nickName",
                  "kind": "field",
                  "label": "Nickname",
                  "style": Object {
                    "flex": 2,
                  },
                },
              ],
              "kind": "flex",
            },
            Object {
              "id": "email",
              "kind": "field",
              "label": "Email",
              "type": "email",
            },
            Object {
              "editable": false,
              "id": "password",
              "kind": "field",
              "label": "Password",
              "type": "password",
            },
            Object {
              "editable": false,
              "id": "password2",
              "kind": "field",
              "label": "Confirm Password",
              "type": "password",
            },
          ],
          "kind": "group",
          "label": "Personal Info",
        },
      ]
    }
    loading={false}
    values={
      Object {
        "_id": "123",
        "email": "mooseman@nowhere.com",
        "firstName": "Moose",
        "lastName": "Man",
        "superAdmin": true,
      }
    }
  />
</div>
`;
